File: coding-labs/retry-timeout-jitter/python/src/retry_policy.py
from dataclasses import dataclass

class RetryableFailure(Exception): pass
class PermanentFailure(Exception): pass

@dataclass(frozen=True)
class Result:
    success: bool
    value: object
    failure: Exception | None
    attempts: int

class RetryPolicy:
    def __init__(self, max_attempts, max_elapsed_ms, base_delay_ms, max_delay_ms, now_ms, sleep, random_unit):
        self.max_attempts, self.max_elapsed_ms = max_attempts, max_elapsed_ms
        self.base_delay_ms, self.max_delay_ms = base_delay_ms, max_delay_ms
        self.now_ms, self.sleep, self.random_unit = now_ms, sleep, random_unit

    def execute(self, operation):
        started, last = self.now_ms(), None
        for attempt in range(1, self.max_attempts + 1):
            try: return Result(True, operation(), None, attempt)
            except PermanentFailure as failure: return Result(False, None, failure, attempt)
            except RetryableFailure as failure:
                last = failure
                if attempt == self.max_attempts: break
                cap = min(self.max_delay_ms, self.base_delay_ms * (2 ** (attempt - 1)))
                delay = int(cap * self.random_unit())
                if self.now_ms() - started + delay > self.max_elapsed_ms: break
                self.sleep(delay)
        return Result(False, None, last, attempt if last else 0)

File: coding-labs/retry-timeout-jitter/java/src/main/java/RetryPolicy.java
import java.util.function.DoubleSupplier;
import java.util.function.LongConsumer;
import java.util.function.LongSupplier;

public final class RetryPolicy<T> {
  @FunctionalInterface public interface Operation<T> { T run(); }
  public static class RetryableFailure extends RuntimeException { public RetryableFailure(String message) { super(message); } }
  public static class PermanentFailure extends RuntimeException { public PermanentFailure(String message) { super(message); } }
  public record Result<T>(boolean success, T value, RuntimeException failure, int attempts) {}
  private final int maxAttempts; private final long maxElapsed; private final long baseDelay; private final long maxDelay; private final LongSupplier clock; private final LongConsumer sleeper; private final DoubleSupplier random;
  public RetryPolicy(int maxAttempts, long maxElapsed, long baseDelay, long maxDelay, LongSupplier clock, LongConsumer sleeper, DoubleSupplier random) { this.maxAttempts = maxAttempts; this.maxElapsed = maxElapsed; this.baseDelay = baseDelay; this.maxDelay = maxDelay; this.clock = clock; this.sleeper = sleeper; this.random = random; }
  public Result<T> execute(Operation<T> operation) {
    long started = clock.getAsLong(); RuntimeException last = null; int attempts = 0;
    for (int attempt = 1; attempt <= maxAttempts; attempt++) {
      attempts = attempt;
      try { return new Result<>(true, operation.run(), null, attempt); }
      catch (PermanentFailure failure) { return new Result<>(false, null, failure, attempt); }
      catch (RetryableFailure failure) { last = failure; if (attempt == maxAttempts) break; long cap = Math.min(maxDelay, baseDelay * (1L << Math.min(attempt - 1, 30))); long delay = (long) Math.floor(cap * random.getAsDouble()); if (clock.getAsLong() - started + delay > maxElapsed) break; sleeper.accept(delay); }
    }
    return new Result<>(false, null, last, attempts);
  }
}
